src/omv/ports/stm32/jpeg.c

```diff
@@ -0,0 +1,707 @@
+    memset(dst->data + sizeof(uint32_t) + sizeof(JPEG_APP0), 0, app0_padding_size - sizeof(uint16_t)); // data
+
+    // Clean trailing data after 0xFFD9 at the end of the jpeg byte stream.
+    dst->size = jpeg_clean_trailing_bytes(dst->size, dst->data);
+
+    #if (TIME_JPEG == 1)
+    printf("compress time: %u ms\n", mp_hal_ticks_ms() - start);
+    #endif
+
+exit_cleanup:
+
+    // Cleanup jpeg state.
+    HAL_JPEG_Abort(&JPEG_state.jpeg_descr);
+    HAL_JPEG_UnRegisterDataReadyCallback(&JPEG_state.jpeg_descr);
+    HAL_JPEG_UnRegisterGetDataCallback(&JPEG_state.jpeg_descr);
+
+    fb_free(); // mcu_row_buffer (after DMA is aborted)
+
+    return jpeg_overflow;
+}
+
+static void jpeg_decompress_data_ready_abort(JPEG_HandleTypeDef *hjpeg, uint8_t *pDataOut, uint32_t OutDataLength) {
+    HAL_JPEG_Abort(hjpeg);
+}
+
+static void jpeg_decompress_get_data(JPEG_HandleTypeDef *hjpeg, uint32_t NbDecodedData) {
+    JPEG_state.jpeg_descr.pJpegInBuffPtr += NbDecodedData;
+    JPEG_state.in_data_len -= NbDecodedData;
+    HAL_JPEG_ConfigInputBuffer(&JPEG_state.jpeg_descr, JPEG_state.jpeg_descr.pJpegInBuffPtr,
+                               IM_MIN(JPEG_state.in_data_len, JPEG_MAX_MDMA_BLOCK_SIZE));
+}
+
+static void jpeg_decompress_data_ready(JPEG_HandleTypeDef *hjpeg, uint8_t *pDataOut, uint32_t OutDataLength) {
+    // We have received this much data.
+    JPEG_state.out_data_len += OutDataLength;
+
+    int remaining = JPEG_state.out_data_len_max - JPEG_state.out_data_len;
+
+    if (!remaining) {
+        HAL_JPEG_Pause(hjpeg, JPEG_PAUSE_RESUME_OUTPUT);
+        JPEG_state.out_data_len = 0;
+        JPEG_state.output_paused = true;
+    } else {
+        // We are ok to receive more data.
+        HAL_JPEG_ConfigOutputBuffer(hjpeg, pDataOut + OutDataLength, IM_MIN(remaining, JPEG_OUTPUT_CHUNK_SIZE));
+    }
+}
+
+void jpeg_decompress(image_t *dst, image_t *src) {
+    #if (TIME_JPEG == 1)
+    mp_uint_t start = mp_hal_ticks_ms();
+    #endif
+
+    // Verify the jpeg image is not a non-baseline jpeg image and check that is has
+    // valid headers up to the start-of-scan header (which cannot be trivially walked).
+    if (!jpeg_is_valid(src)) {
+        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("Non-Baseline JPEGs are not supported."));
+    }
+
+    // Ensure src data is cache algined for MDMA and a multiple of fifo input size in bytes.
+    JPEG_state.in_data_len = src->size;
+    uint32_t diff = JPEG_state.in_data_len % JPEG_INPUT_FIFO_BYTES;
+
+    if (diff) {
+        JPEG_state.in_data_len += JPEG_INPUT_FIFO_BYTES - diff;
+    }
+
+    // Copy to cache aligned buffer.
+    JPEG_state.jpeg_descr.pJpegInBuffPtr = fb_alloc(JPEG_state.in_data_len, FB_ALLOC_CACHE_ALIGN);
+    memcpy(JPEG_state.jpeg_descr.pJpegInBuffPtr, src->data, src->size);
+
+    // Zero remaining.
+    memset(JPEG_state.jpeg_descr.pJpegInBuffPtr + src->size, 0, JPEG_state.in_data_len - src->size);
```

Comment:
There's room for a minor optimization here, maybe the source buffer is already cache-aligned ?

Reply:
Added it. This is actually a huge speedup if aligned if the JPEG is large.

Reply:
@iabdalkader Making this change destroyed JPEG performance. This is because this keeps the JPEG buffer in SDRAM when MDMA is accessing it. Sending another PR to remove this change.

Reply:
Mmm, on further testing, this doesn't seem to be the culprit. However, JPEG performance is generally less than what I measured before I made many of the requested changes. Will need to figure out what is costing so much speed.

Something else though is causing a vary serious slowdown when trying to play MJPEG files.

Reply:
Mmm, false alarm. So, when the jpeg decoder gets stuck it just partially decodes an image and you busy wait for 1 second on the tiemout.

Reply:
https://github.com/openmv/openmv/pull/2093 - Would have saved hours of debugging.

Reply:
I missed this, all good now, right ?

Reply:
Yeah.